src/Gaffer/RandomChoice.cpp

```diff
@@ -0,0 +1,355 @@
+}
+
+const ValuePlug *RandomChoice::choicesPlug() const
+{
+	return getChild<ValuePlug>( g_firstPlugIndex + 2 );
+}
+
+FloatVectorDataPlug *RandomChoice::choicesWeightsPlug()
+{
+	return choicesPlug()->getChild<FloatVectorDataPlug>( 0 );
+}
+
+const FloatVectorDataPlug *RandomChoice::choicesWeightsPlug() const
+{
+	return choicesPlug()->getChild<FloatVectorDataPlug>( 0 );
+}
+
+ValuePlug *RandomChoice::outPlugInternal()
+{
+	return getChild<ValuePlug>( g_outPlugName );
+}
+
+const ValuePlug *RandomChoice::outPlugInternal() const
+{
+	return getChild<ValuePlug>( g_outPlugName );
+}
+
+void RandomChoice::affects( const Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	ComputeNode::affects( input, outputs );
+
+	if(
+		input == seedPlug() ||
+		input == seedVariablePlug() ||
+		input == choicesWeightsPlug() ||
+		input == choicesValuesPlug()
+	)
+	{
+		if( const Plug *p = outPlug() )
+		{
+			if( p->children().size() )
+			{
+				for( const auto &cp : Plug::Range( *p ) )
+				{
+					outputs.push_back( cp.get() );
+				}
+			}
+			else
+			{
+				outputs.push_back( p );
+			}
+		}
+	}
+}
+
+void RandomChoice::hash( const ValuePlug *output, const Context *context, IECore::MurmurHash &h ) const
+{
+	ComputeNode::hash( output, context, h );
+
+	if( output == outPlug() || output->parent() == outPlug() )
+	{
+		seedPlug()->hash( h );
+		const string seedVariable = seedVariablePlug()->getValue();
+		if( !seedVariable.empty() )
+		{
+			h.append( context->variableHash( seedVariable ) );
+		}
+		choicesWeightsPlug()->hash( h );
+		choicesValuesPlug()->hash( h );
+		h.append( output->defaultHash() );
+	}
+}
+
+void RandomChoice::compute( ValuePlug *output, const Context *context ) const
+{
+	if( output == outPlug() || output->parent() == outPlug() )
+	{
+		ConstFloatVectorDataPtr weightsData = choicesWeightsPlug()->getValue();
+		const vector<float> &weights = weightsData->readable();
+
+		size_t seed = seedPlug()->getValue();
+		const string seedVariable = seedVariablePlug()->getValue();
+		if( !seedVariable.empty() )
+		{
+			/// \todo See comments in `Random::computeSeed()`.
+			IECore::DataPtr contextData = context->getAsData( seedVariable, nullptr );
+			if( contextData )
+			{
+				const MurmurHash hash = contextData->Object::hash();
+				seed += hash.h1();
+			}
+		}
+		Imath::Rand48 random( seed );
+
+		dispatchPlugFunction(
+			outPlug(),
+			[this, output, &weights, &random] ( auto plug ) {
+
+				using PlugType = remove_const_t<remove_pointer_t<decltype( plug )>>;
+				using ValueType = typename PlugType::ValueType;
+
+				auto choicesData = this->choicesValuesPlug<ValuesPlugType<ValueType>>()->getValue();
+				const auto &choices = choicesData->readable();
+				if( !choices.size() )
+				{
+					output->setToDefault();
+					return;
+				}
+
+				// We might have been provided with too few weights or too many. We
+				// try to be helpful by using defaults for missing weights, and ignoring
+				// any extras.
```

Comment:
Hmm.  I know I sometimes prefer halting with an exception a bit too much, but is there any way to enter this mismatch state using the UI?  If you can only get into it using scripting, then halting with an exception to let someone know something has gone wrong could be useful?  Failing to provide enough the right number of weights, and getting a distribution that is statistically different to what you expect could be pretty hard to debug.

Reply:
I don't think you could get into that state using the VectorDataWidget directly, but it would be pretty easy to do using an expression driving `choices.values`. I think you make a good point though. Some options :

1. Always error for mismatches.
2. As above, except if `len( choices.weights ) == 0`, when we use uniform weights. In this case it seems safe to assume the user doesn't care about weights and will be grateful for the default behaviour?
3. Always error for mismatches, but have a `useWeights` toggle that can be turned off if you don't want to have to provide weights.

I think I could be talked into any of them, but I do like the simplicity of 1, and it's not like it would be hard to provide a list of uniform weights from the same expression that provided the list of values.

Reply:
Implemented 1 in https://github.com/GafferHQ/gaffer/pull/4548/commits/ff35810bc69a089155c1a72c7b30ad1829f05d8c